sympy/solvers/solveset.py

```diff
@@ -3665,10 +3665,11 @@ def nonlinsolve(system, *symbols):
         return _solveset_work(system, symbols)
 
     # main code of def nonlinsolve() starts from here
+    nonconstant = [eqn for eqn in map(_sympify, system) if eqn.free_symbols & set(symbols)]
     polys, polys_expr, nonpolys, denominators = _separate_poly_nonpoly(
-        system, symbols)
+        nonconstant, symbols)
 
-    if len(symbols) == len(polys):
+    if len(system) == len(polys):
```

Comment:
What about just `if not nonpolys`?

Reply:
It would be better to solve all the poly equations and substitute those solutions into the remaining equations before entering into substitution.

Reply:
Yes, that makes sense. Are there any gotchas I need to be careful about while substituting?